gapir/client/connection.go

```diff
@@ -0,0 +1,262 @@
+
+func NewPayload(stackSize, volatileMemorySize uint32, constants []byte, resources ResourceInfoList, opcodes []byte) *Payload {
+	return (*Payload)(&replaysrv.Payload{
+		StackSize:          stackSize,
+		VolatileMemorySize: volatileMemorySize,
+		Constants:          constants,
+		Resources:          ([]*replaysrv.ResourceInfo)(resources),
+		Opcodes:            opcodes,
+	})
+}
+
+func (p *Payload) ToProto() replaysrv.Payload {
+	return (replaysrv.Payload)(*p)
+}
+
+type Resources struct {
+	replaysrv.Resources
+}
+
+type ResourceRequest struct {
+	replaysrv.ResourceRequest
+}
+
+type CrashDump struct {
+	replaysrv.CrashDump
+}
+
+type PostData struct {
+	replaysrv.PostData
+}
+
+type Notification struct {
+	replaysrv.Notification
+}
+
+type Connection struct {
+	conn       *grpc.ClientConn
+	servClient replaysrv.GapirClient
+	stream     replaysrv.Gapir_ReplayClient
+	authToken  auth.Token
+}
+
+func newConnection(addr string, authToken auth.Token, timeout time.Duration) (*Connection, error) {
+	conn, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(timeout))
+	if err != nil {
+		return nil, err
+	}
+	s := replaysrv.NewGapirClient(conn)
+	return &Connection{conn: conn, servClient: s, authToken: authToken}, nil
+}
+
+func (c *Connection) Close() {
+	if c.stream != nil {
+		c.stream.CloseSend()
+	}
+	if c.conn != nil {
+		c.conn.Close()
+	}
+	c.conn = nil
+	c.servClient = nil
+	c.stream = nil
+}
+
+func (c *Connection) Ping(ctx context.Context) error {
+	if c.servClient == nil {
+		return log.Err(ctx, nil, "Gapir not connected")
+	}
+	ctx = c.attachAuthToken(ctx)
+	r, err := c.servClient.Ping(ctx, &replaysrv.PingRequest{})
+	if err != nil {
+		return log.Err(ctx, err, "Sending ping")
+	}
+	if r == nil {
+		return log.Err(ctx, nil, "No response for ping")
+	}
+	return nil
+}
+
+func (c *Connection) Shutdown(ctx context.Context) error {
+	if c.servClient == nil {
+		return log.Err(ctx, nil, "Gapir not connected")
+	}
+	ctx = c.attachAuthToken(ctx)
+	_, err := c.servClient.Shutdown(ctx, &replaysrv.ShutdownRequest{})
+	if err != nil {
+		return log.Err(ctx, err, "Sending shutdown request")
+	}
+	return nil
+}
+
+func (c *Connection) SendResources(ctx context.Context, resources []byte) error {
+	if c.conn == nil || c.servClient == nil {
+		return log.Err(ctx, nil, "Gapir not connected")
+	}
+	if c.stream == nil {
+		return log.Err(ctx, nil, "Replay communication not initiated")
+	}
+	resReq := replaysrv.ReplayRequest{
+		Req: &replaysrv.ReplayRequest_Resources{
+			Resources: &replaysrv.Resources{Data: resources},
+		},
+	}
+	if err := c.stream.Send(&resReq); err != nil {
+		return log.Err(ctx, err, "Sneding resources")
+	}
+	return nil
+}
+
+func (c *Connection) HandleReplayCommunication(
+	ctx context.Context,
+	replayID string,
+	payload Payload,
+	handleResourceRequest func(context.Context, *ResourceRequest, *Connection) error,
+	handleCrashDump func(context.Context, *CrashDump, *Connection) error,
```

Comment:
Consider putting the `Connection` into the context. This'll simplify the callbacks, and let you fetch it much deeper down in a call chain.